crates/sncast/src/starknet_commands/declare.rs

```diff
@@ -54,79 +55,105 @@ impl_payable_transaction!(Declare, token_not_supported_for_declaration,
-                ErrorData::new(declare.contract),
-            ))?;
-
-    let contract_definition: SierraClass = serde_json::from_str(&contract_artifacts.sierra)
-        .context("Failed to parse sierra artifact")?;
-    let casm_contract_definition: CompiledClass =
-        serde_json::from_str(&contract_artifacts.casm).context("Failed to parse casm artifact")?;
+type SendDeclarationError<'client> = AccountError<<SingleOwnerAccount<&'client JsonRpcClient<HttpTransport>, LocalWallet> as starknet::accounts::Account>::SignError>;
 
-    let casm_class_hash = casm_contract_definition
-        .class_hash()
-        .map_err(anyhow::Error::from)?;
+async fn send_declaration<'client>(
+    contract: CompiledContract,
+    account: &'client SingleOwnerAccount<&'client JsonRpcClient<HttpTransport>, LocalWallet>,
+    nonce: Option<FieldElement>,
+    fee_settings: FeeSettings,
+) -> Result<DeclareTransactionResult, SendDeclarationError<'client>> {
+    let CompiledContract { class, hash } = contract;
 
-    let declared = match fee_settings {
+    match fee_settings {
         FeeSettings::Eth { max_fee } => {
-            let declaration = account.declare_v2(
-                Arc::new(contract_definition.flatten().map_err(anyhow::Error::from)?),
-                casm_class_hash,
-            );
+            let declaration = account.declare_v2(Arc::new(class), hash);
 
             let declaration = apply_optional(declaration, max_fee, DeclarationV2::max_fee);
-            let declaration = apply_optional(declaration, declare.nonce, DeclarationV2::nonce);
+            let declaration = apply_optional(declaration, nonce, DeclarationV2::nonce);
 
             declaration.send().await
         }
+
         FeeSettings::Strk {
             max_gas,
             max_gas_unit_price,
         } => {
-            let declaration = account.declare_v3(
-                Arc::new(contract_definition.flatten().map_err(anyhow::Error::from)?),
-                casm_class_hash,
-            );
+            let declaration = account.declare_v3(Arc::new(class), hash);
 
             let declaration = apply_optional(declaration, max_gas, DeclarationV3::gas);
             let declaration =
                 apply_optional(declaration, max_gas_unit_price, DeclarationV3::gas_price);
-            let declaration = apply_optional(declaration, declare.nonce, DeclarationV3::nonce);
+            let declaration = apply_optional(declaration, nonce, DeclarationV3::nonce);
 
             declaration.send().await
         }
-    };
-
-    match declared {
-        Ok(result) => handle_wait_for_tx(
-            account.provider(),
-            result.transaction_hash,
-            DeclareResponse {
-                class_hash: Felt(result.class_hash),
-                transaction_hash: Felt(result.transaction_hash),
-            },
-            wait_config,
-        )
-        .await
-        .map_err(StarknetCommandError::from),
+    }
+}
+
+async fn handle_declaration<'client>(
```

Comment:
If this is not reused, i don't think it provides much readability in this case

Reply:
I side with @Arcticae here. You need to create type aliases and introduce lifetimes for `send_declaration` and `handle_declaration`, so what you gain by separating logic into separate functions, you lose due to their weird function signatures. I would remove these two functions; the rest is fine for me.